src/buffer/out/Row.cpp

```diff
@@ -191,8 +347,331 @@ OutputCellIterator ROW::WriteCells(OutputCellIterator it, const til::CoordType i
+
+    if (chExtEndNew != chExtEnd)
+    {
+        _resizeChars(colExtEnd, chExtBeg, chExtEnd, chExtEndNew);
+    }
+
+    // Add leading/trailing whitespace and copy chars
+    {
+        auto it = _chars.begin() + chExtBeg;
+        it = fill_n_small(it, leadingSpaces, L' ');
+        it = copy_n_small(chars.begin(), chars.size(), it);
+        it = fill_n_small(it, trailingSpaces, L' ');
+    }
+    // Update char offsets with leading/trailing whitespace and the chars columns.
+    {
+        auto chPos = chExtBeg;
+        auto it = _charOffsets.begin() + colExtBeg;
+
+        it = iota_n_mut(it, leadingSpaces, chPos);
+
+        *it++ = chPos;
+        it = fill_small(it, _charOffsets.begin() + colEnd, gsl::narrow_cast<uint16_t>(chPos | CharOffsetsTrailer));
+        chPos = gsl::narrow_cast<uint16_t>(chPos + chars.size());
+
+        it = iota_n_mut(it, trailingSpaces, chPos);
+    }
+}
+
+const til::small_rle<TextAttribute, uint16_t, 1>& ROW::Attributes() const noexcept
+{
+    return _attr;
+}
+
+TextAttribute ROW::GetAttrByColumn(const til::CoordType column) const
+{
+    return _attr.at(_clampedUint16(column));
+}
+
+std::vector<uint16_t> ROW::GetHyperlinks() const
+{
+    std::vector<uint16_t> ids;
+    for (const auto& run : _attr.runs())
+    {
+        if (run.value.IsHyperlink())
+        {
+            ids.emplace_back(run.value.GetHyperlinkId());
+        }
+    }
+    return ids;
+}
+
+uint16_t ROW::size() const noexcept
+{
+    return _columnCount;
+}
+
+til::CoordType ROW::MeasureLeft() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = beg;
+
+    for (; it != end; ++it)
+    {
+        if (*it != L' ')
+        {
+            break;
+        }
+    }
+
+    return gsl::narrow_cast<til::CoordType>(it - beg);
+}
+
+til::CoordType ROW::MeasureRight() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = end;
+
+    for (; it != beg; --it)
+    {
+        // it[-1] is safe as `it` is always greater than `beg` (loop invariant).
+        if (til::at(it, -1) != L' ')
+        {
+            break;
+        }
+    }
+
+    // We're supposed to return the measurement in cells and not characters
+    // and therefore simply calculating `it - beg` would be wrong.
+    //
+    // An example: The row is 10 cells wide and `it` points to the second character.
+    // `it - beg` would return 1, but it's possible it's actually 1 wide glyph and 8 whitespace.
+    return gsl::narrow_cast<til::CoordType>(_columnCount - (end - it));
+}
+
+bool ROW::ContainsText() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = beg;
+
+    for (; it != end; ++it)
+    {
+        if (*it != L' ')
+        {
+            return true;
+        }
+    }
+
+    return false;
+}

```

Comment:
(more of a fun fact than anything) I guess you could use `MeasureLeft` or `MeasureRight` if you wanted to simplify this code.

Reply:
those methods need to die

Reply:
Technically none of the 3 functions is needed anywhere. `MeasureLeft()` and `ContainsText()` are both only used in unit tests and `MeasureRight()`'s sole purpose is to remeasure the end of the buffer in `TextBuffer::Reflow()` (via `GetLastNonSpaceCharacter()`), which is kind of... "eh", because it shouldn't have to do that in the first place. 😅